pkg/controllers/management/restrictedadminrbac/fleet.go

```diff
@@ -50,42 +51,72 @@ func (r *rbaccontroller) ensureRestricedAdminForFleet(key string, obj *v3.Global
 }
 
 func (r *rbaccontroller) ensureRolebinding(namespace string, subject k8srbac.Subject, grb *v3.GlobalRoleBinding) error {
-	rbName := fmt.Sprintf("%s-fleetworkspace-%s", grb.Name, rbac.RestrictedAdminClusterRoleBinding)
-	rb := &k8srbac.RoleBinding{
-		ObjectMeta: metav1.ObjectMeta{
-			Name:      rbName,
-			Namespace: namespace,
-			Labels:    map[string]string{rbac.RestrictedAdminClusterRoleBinding: "true"},
-			OwnerReferences: []metav1.OwnerReference{
-				{
-					APIVersion: grb.TypeMeta.APIVersion,
-					Kind:       grb.TypeMeta.Kind,
-					UID:        grb.UID,
-					Name:       grb.Name,
-				},
-			},
-		},
-		RoleRef: k8srbac.RoleRef{
-			Name: "fleetworkspace-admin",
-			Kind: "ClusterRole",
-		},
-		Subjects: []k8srbac.Subject{
-			subject,
+	name := fmt.Sprintf("%s-fleetworkspace-%s", grb.Name, rbac.RestrictedAdminClusterRoleBinding)
+	ownerRefs := []metav1.OwnerReference{
+		{
+			APIVersion: grb.TypeMeta.APIVersion,
+			Kind:       grb.TypeMeta.Kind,
+			UID:        grb.UID,
+			Name:       grb.Name,
 		},
 	}
-	_, err := r.rbLister.Get(namespace, rbName)
+	roleRef := k8srbac.RoleRef{
+		Name: "fleetworkspace-admin",
+		Kind: "ClusterRole",
+	}
+	subjects := []k8srbac.Subject{
+		subject,
+	}
+
+	rb, err := r.rbLister.Get(namespace, name)
 	if err != nil && !k8serrors.IsNotFound(err) {
+		// list call failed for unknown reason, give up
 		return err
-	} else if k8serrors.IsNotFound(err) {
-		_, err = r.roleBindings.Create(rb)
+	}
+
+	// role binding not found, create it
+	if err != nil && k8serrors.IsNotFound(err) {
+		_, err = r.roleBindings.Create(&k8srbac.RoleBinding{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:            name,
+				Namespace:       namespace,
+				Labels:          map[string]string{rbac.RestrictedAdminClusterRoleBinding: "true"},
+				OwnerReferences: ownerRefs,
+			},
+			RoleRef:  roleRef,
+			Subjects: subjects,
+		})
 		if err != nil && !k8serrors.IsAlreadyExists(err) {
 			return err
 		}
-	} else {
-		_, err := r.roleBindings.Update(rb)
-		if err != nil {
-			return err
-		}
+		return nil
```

Comment:
Can we return nil here in the case that we got an already exists error above? I think that in the case that we have a binding which already exists we need to go through the below integrity check, right?

Reply:
That was the previous behavior, and from my perspective it's reasonable.

Yes, there is a small chance that between the Get call and the Create call somebody else has created a RoleBinding with the same name that also happens to be incorrect. If it's Rancher code, then that would be a bug that should be fixed rather than worked around here. If it's a user, I'd let that be reconciled on the next handler call rather than listing -> creating -> listing again -> potentially updating.

Does that make sense to you?

Reply:
> That was the previous behavior

I can resolve this on that basis, I don't think it makes sense to require fixing other issues as part of this PR.

That being said, I do not agree with the logic that this behavior is desired. The binding that this code is creating is managed by Rancher, IMO any corruption/change of that binding should be corrected as quickly as possible. Relying on the next handler call to fix it leaves an indeterminate time between when the corrupt binding was created and when it will be fixed (this could be called soon if a relatedResource change, or could wait a long time between now and the next adjustment).

Reply:
I agree this is best settled in a separate PR.

Thanks also for the insights, I see where you are coming from.

> The binding that this code is creating is managed by Rancher, IMO any corruption/change of that binding should be corrected as quickly as possible.

Curiosity: would it be possible or even good to have similar "protected objects" actually completely guarded by Rancher's Webhook, so that nothing but Rancher can actually touch them? Such a design would make this whole "drift detection" logic useless (after, possibly, one last round of cleanup).